src/main/java/org/jabref/logic/importer/util/MetaDataParser.java

```diff
@@ -49,61 +51,55 @@ public MetaData parse(MetaData metaData, Map<String, String> data, Character key
         List<String> defaultCiteKeyPattern = new ArrayList<>();
         Map<EntryType, List<String>> nonDefaultCiteKeyPatterns = new HashMap<>();
 
-        for (Map.Entry<String, String> entry : data.entrySet()) {
+        // process GROUPSTREE and GROUPSTREE_LEGACY at the very end (otherwise it can happen that not all dependent data are set)
+        Stream<Map.Entry<String, String>> entrySetStream = data.entrySet().stream().filter(entry -> !entry.getKey().equals(MetaData.GROUPSTREE) && !entry.getKey().equals(MetaData.GROUPSTREE_LEGACY));
+        Stream<Map.Entry<String, String>> entrySetStreamTail = data.entrySet().stream().filter(entry -> entry.getKey().equals(MetaData.GROUPSTREE) || entry.getKey().equals(MetaData.GROUPSTREE_LEGACY));
+        entrySetStream = Stream.concat(entrySetStream, entrySetStreamTail);
+
+        Iterator<Map.Entry<String, String>> entryIterator = entrySetStream.iterator();
+
+        while (entryIterator.hasNext()) {
```

Comment:
A more "streamy" way is using the `stream.foreach(...)` method.

Reply:
In the loop exceptions may be thrown. It's slightly complicated to throw them back within a lambda expression (at least for me).

Reply:
ok, good argument. I think first using `collect` to get an ordinary list and then using foreach is still slightly better.